pkg/cmd/admin/migrate/images/images.go

```diff
@@ -0,0 +1,488 @@
+
+  # Migrate from a service IP to an internal service DNS name for all deployment configs and builds
+  %[1]s 172.30.1.54/*=registry.openshift.svc.cluster.local/* --include=buildconfigs,deploymentconfigs`
+)
+
+type MigrateImageReferenceOptions struct {
+	migrate.ResourceOptions
+
+	Client   client.Interface
+	Mappings ImageReferenceMappings
+}
+
+// NewCmdMigrateImageReferences implements a MigrateImages command
+func NewCmdMigrateImageReferences(name, fullName string, f *clientcmd.Factory, in io.Reader, out, errout io.Writer) *cobra.Command {
+	options := &MigrateImageReferenceOptions{
+		ResourceOptions: migrate.ResourceOptions{
+			In:      in,
+			Out:     out,
+			ErrOut:  errout,
+			Include: []string{"imagestream", "image", "secrets"},
+		},
+	}
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s REGISTRY/NAME=REGISTRY/NAME [...]", name),
+		Short:   "Update embedded Docker image references",
+		Long:    internalMigrateImagesLong,
+		Example: fmt.Sprintf(internalMigrateImagesExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Complete(f, cmd, args))
+			kcmdutil.CheckErr(options.Validate())
+			if err := options.Run(); err != nil {
+				// TODO: move met to kcmdutil
+				if err == cmdutil.ErrExit {
+					os.Exit(1)
+				}
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+	options.ResourceOptions.Bind(cmd)
+
+	return cmd
+}
+
+func (o *MigrateImageReferenceOptions) Complete(f *clientcmd.Factory, c *cobra.Command, args []string) error {
+	var remainingArgs []string
+	for _, s := range args {
+		if !strings.Contains(s, "=") {
+			remainingArgs = append(remainingArgs, s)
+			continue
+		}
+		mapping, err := ParseMapping(s)
+		if err != nil {
+			return err
+		}
+		o.Mappings = append(o.Mappings, mapping)
+	}
+
+	o.ResourceOptions.SaveFn = o.save
+	if err := o.ResourceOptions.Complete(f, c, remainingArgs); err != nil {
+		return err
+	}
+
+	osclient, _, err := f.Clients()
+	if err != nil {
+		return err
+	}
+	o.Client = osclient
+
+	return nil
+}
+
+func (o *MigrateImageReferenceOptions) Validate() error {
+	if len(o.Mappings) == 0 {
+		return fmt.Errorf("at least one mapping argument must be specified: REGISTRY/NAME=REGISTRY/NAME")
+	}
+	return o.ResourceOptions.Validate()
+}
+
+func (o *MigrateImageReferenceOptions) Run() error {
+	return o.ResourceOptions.Visit(func(info *resource.Info) (migrate.Reporter, error) {
+		return o.transform(info.Object)
+	})
+}
+
+// save invokes the API to alter an object. The reporter passed to this method is the same returned by
+// the migration visitor method (for this type, transformImageReferences). It should return an error
+// if the input type cannot be saved. It returns migrate.ErrRecalculate if migration should be re-run
+// on the provided object.
+func (o *MigrateImageReferenceOptions) save(info *resource.Info, reporter migrate.Reporter) error {
+	switch t := info.Object.(type) {
+	case *imageapi.ImageStream:
+		if reporter.(imageChangeInfo).status {
```

Comment:
If we have any status changes we always have to recalculate anyway (calculating of spec vs status may result in a migration of status being necessary again).